File: 0-notes/job-search/Cracking the Coding Interview/C01ArraysStrings/javascript/1.4-answer.js
// Palindrome Permutation

// TODO:  Given a string, write a function to check if it is a permutation of a palindrome.
// ! A palindrome is a word or phrase that is the same forwards and backwards.
// ! A permutation is a rearrangement of letters.
// ! The palindrome does not need to be limited to just dictionary words.
// ! You can ignore casing and non-letter characters.

function palindromePerm(s){
    let counts = {};
    s = s.toLowerCase();
    let s_array = [];
    for (var x = 0; x < s.length; x++) {
        if(s.charAt(x).match(/[a-z]/i)){
            s_array.push(s.charAt(x));
        }
    }
    console.log(s_array);
    for(let x = 0; x < s_array.length; x++) {
        current_letter = s_array[x];
        if(!counts[current_letter]){
            counts[current_letter] = 1;
        }
        else{
            counts[current_letter] = counts[current_letter]+1;
        }
    };
    let count_doubles = 0;
    for (let key of Object.keys(counts)) {
        if(counts[key] == 1){
            count_doubles++;
        }
      }
    console.log(count_doubles);
    if(count_doubles < 2){
        console.log("palindrome");
        return true;
    }

    console.log("not a palindrome");
    return false;
}

palindromePerm("tacocat");
palindromePerm("racecar");
palindromePerm("Adam");
palindromePerm("Erica");

File: 0-notes/job-search/Cracking the Coding Interview/C01ArraysStrings/python/1.4-answer.py
# Palindrome Permutation

# Given a string, write a function to check if it is a permutation of a palindrome.
# A palindrome is a word or phrase that is the same forwards and backwards.
# A permutation is a rearrangement of letters.
# The palindrome does not need to be limited to just dictionary words.
# You can ignore casing and non-letter characters.
import re
def palindromePerm(string):
    s1 = ""
    for letter in string:
        if letter.isalpha():
            s1 = s1 + letter
    s1 = str(s1.lower())
    print("s1 ", s1)
    # set up a list
    s1_list = set(s1)
    s1_counts = [0] * len(s1_list)
    s1_dict = {}
    for list_item in s1_list:
        s1_dict[list_item] = 0
    # keep count of each letter in this list
    for letter in s1:
        s1_dict[letter] += 1
    length = len(s1)
    for key, value in s1_dict.items():
        print(key, value)
    # with even length:  if all counts even, no odd counts = True
    if length % 2 == 0:
        for key, value in s1_dict.items():
            if value % 2 != 0:
                print(False)
                return False
    else:
        # with odd length:  if all counts even but one 1 count = True
        for key, value in s1_dict.items():
            if value % 2 != 0:
                for key1, value1 in s1_dict.items():
                    if value1 % 2 != 0 and key != key1:
                        print("else", False)
                        return False
    # return true if you made it here
    print(True)
    return True 


palindromePerm("taco cat")
